### PR TITLE
Visualising waypoints 4

### LINKED ISSUE #2598: Visualize waypoints set by the player
You can currently use Shift + Click to set waypoints for your units. There should be a flag (suggesting the one from the rally point) to visualize this.

### LINKED ISSUE #12019: Target lines need an overhaul
Most of the `SetTargetLine` calls are currently done from trait code, which means that when you're queuing commands, the last command overrides all the previous target lines and becomes the only one visible.

A solution would be to instead call `SetTargetLine` when an activity is about to run for the first time.

### PR DESCRIPTION
**Pull request by tovl**
Fixes #2598
Fixes #12019

This is completing the work that @Turupawn started and @forcecore continued in #14384 .
I've rebased it on top of #16481 to complete the general activity rework and added a few modifications of my own:

- Made target lines thicker (Maybe it's because of my high-resolution monitor or my eyes are just bad, but I found the single pixel lines hard to read when planning queued orders)
- Omitted target lines from autotarget attacks and AttackMove/Guard attacks.
- Changed some colors of target lines: 
  - AttackMove and Guard get yellow (because they mix move -> green and attack -> red, yellow seems intuitive)
  - ~~'Special' non-aggressive actions changed from yellow to blue.~~

In general only actions that directly correspond to given orders should have wapoint lines rendered. I may possibly reintroduce special target lines for autotarget, AttackMove/Guard and opportunity fire targets in a future PR. These target lines would be completely seperate from the waypoint lines, not part of the chain and rendered in thinner lines.

### CODE REVIEW

**Comment by pchote** [comment]
> In general only actions that directly correspond to given orders should have wapoint lines rendered. 

IMO the rule of thumb here should be that we show the line for the parent activity, and ignore any children - so AttackMove would show the line to the move target, Guard to to the unit that is being guarded, the various things that use `MoveAdjacentTo` to the center of the actor that is being moved next to, etc.

**Comment by pchote** [comment]
The visual changes are likely going to be controversial, so can be please defer those to a followup to avoid bogging this important PR down with visual bikeshedding?

**Comment by pchote** [comment]
I thought we had an issue about the target lines getting out of sync, which this will fix, but can't find it.

**Comment by dragunoff** [comment]
I agree with pchote that the visual changes are controversial and are best left to follow-ups. The only visual change that I'd suggest here is adding a small square/dot at each node in the path.

**Comment by tovl** [comment]
The color changes here are not about aesthetics though, but rather just basic functionality. The whole point here is to give players a visual reminder of the orders they have queued. I think that at the very least they should be able to distinguish attack, move and attack-move orders. Giving them three different colors is the most straightforward way to do that _without_ getting too deep into aesthetic details.

It then also makes sense to have a fourth color for orders that do not really correspond to either move/attack/attack-move. Which color that is doesn't really concern me as long as it's easily distinguishable; Blue seems as good as any.

**Comment by pchote** [comment]
I like the idea of expanding the color palette, but I think we're going to spend some time hashing out the specific scheme - for example, the previous convention of gold for repair/guard/heal was based on the gold wrench / yellow enter icon for engineers, while we have a blue cursor for other generic abilities.

Something about the wide lines with missing end dots really feels wrong to me, so I'd also like to iterate on this further too. I don't think going back to the old lines is necessarily the best option though - it would be nice if we could e.g. show striped red/green lines for attackmove and i'm not sure that would work with only a 1px line.

But again, these cosmetic things really IMO shouldn't drag down the core work here, which we need to fix the target line regressions introduced in the last release.

**Comment by matjaeck** [comment]
> The color changes here are not about aesthetics though, but rather just basic functionality. The whole point here is to give players a visual reminder of the orders they have queued. I think that at the very least they should be able to distinguish attack, move and attack-move orders. Giving them three different colors is the most straightforward way to do that without getting too deep into aesthetic details.

It is not useful though for colorblind people unless you allow users to define the line style themselves. Specifically green/red is a bad combination when you have no other indicators.

> The only visual change that I'd suggest here is adding a small square/dot at each node in the path.

> it would be nice if we could e.g. show striped red/green lines for attackmove and i'm not sure that would work with only a 1px line.

I'm not sure if it would look good, but we could try to show an overlay with the command bar icon that represents the queued activity at the line endings.

**Comment by pchote** [comment]
I created #16616 with additional context and my own thoughts on the line visuals for discussion.

**Comment by pchote** [comment]
It would be good if we can still get the `TargetLineNodes` part of this PR in for the next release (to fix the target line mismatch bugs that exist on the current release), but in the interests of time we should probably defer the queued renderering and other visual changes to Next + 1.

**Comment by tovl** [comment]
We should get #16481 merged first before it makes sense to continue working on this.

**Comment by pchote** [comment]
#16481 has been merged.

**Comment by Turupawn** [comment]
Just letting you guys know that I'm rooting very hard for you on this 🏄‍♀️
Great to see how it's progressing

**Comment by tovl** [comment]
Rebased.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 1]
```diff
@@ -45,6 +45,9 @@
     <PackageReference Include="StyleCop.Analyzers" Version="1.1.118" />
     <AdditionalFiles Include="../stylecop.json" />
   </ItemGroup>
+  <ItemGroup>
+    <Compile Remove="Activities\Air\HeliAttack.cs" />
+  </ItemGroup>
```
Comment (teinarss) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Im guessing this should not be here?

**Review by teinarss — DISMISSED** [review]

**Comment by tovl** [comment]
Removed the unnecessary lines.

**Comment by teinarss** [comment]
No lines when sending units for repair. Not sure if there is a connection to #16721

**Comment by tovl** [comment]
> No lines when sending units for repair. 

Fixed and rebased. Note that move commands queued after repair are still executed in the wrong order, but this is a separate issue that will hopefully be fixed by #16721.

File: OpenRA.Game/Activities/Activity.cs [hunk 2]
```diff
@@ -12,12 +12,29 @@
...
 using System.Collections.Generic;
 using System.Linq;
+using OpenRA.Primitives;
 using OpenRA.Traits;
 
 namespace OpenRA.Activities
 {
 	public enum ActivityState { Queued, Active, Canceling, Done }
 
+	public struct TargetLineNode
```
Comment (teinarss) [inline OpenRA.Game/Activities/Activity.cs]: Not sure if this should be a struct, its size is 88 bytes.  See https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/choosing-between-class-and-struct

**Comment by teinarss** [comment]
Look at the apache in this gif, still showing a line to the helipad after being repaired.
queuing2

File: OpenRA.Game/Activities/Activity.cs (hunk 2 above)
Comment (tovl) [reply OpenRA.Game/Activities/Activity.cs]: Converted to class.

**Comment by tovl** [comment]
> still showing a line to the helipad after being repaired.

Fixed.

**Review by teinarss — DISMISSED** [review]
LGTM

**Comment by tovl** [comment]
Rebased.

File: OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs [hunk 3]
```diff
@@ -62,65 +43,61 @@ void INotifySelected.Selected(Actor a)
...
 		IEnumerable<IRenderable> IRenderAboveShroudWhenSelected.RenderAboveShroud(Actor self, WorldRenderer wr)
 		{
-			var force = Game.GetModifierKeys().HasModifier(Modifiers.Alt);
-			if ((lifetime <= 0 || --lifetime <= 0) && !force)
+			if (self.Owner != self.World.LocalPlayer)
 				yield break;
 
-			if (!(force || Game.Settings.Game.DrawTargetLine))
+			// shift is "force" too, players want to see the lines when in waypoint mode.
+			bool force = Game.GetModifierKeys().HasModifier(Modifiers.Alt)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs]: `var`? Also please start comments with a capital letter: `Shift`

File: OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs [hunk 4]
```diff
@@ -62,65 +43,61 @@ void INotifySelected.Selected(Actor a)
...
+			if ((lifetime <= 0 || --lifetime <= 0) && !force)
 				yield break;
 
-			if (targets == null || targets.Count == 0)
+			if (!(force || Game.Settings.Game.DrawTargetLine))
 				yield break;
 
-			foreach (var target in targets)
+			WPos prev = self.CenterPosition;
+			Activity a = self.CurrentActivity;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs]: `var`?

File: OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs [hunk 5]
```diff
@@ -62,65 +43,61 @@ void INotifySelected.Selected(Actor a)
...
-			if ((lifetime <= 0 || --lifetime <= 0) && !force)
+			if (self.Owner != self.World.LocalPlayer)
 				yield break;
 
-			if (!(force || Game.Settings.Game.DrawTargetLine))
+			// shift is "force" too, players want to see the lines when in waypoint mode.
+			bool force = Game.GetModifierKeys().HasModifier(Modifiers.Alt)
+			          || Game.GetModifierKeys().HasModifier(Modifiers.Shift);
+
+			if ((lifetime <= 0 || --lifetime <= 0) && !force)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs]: Can't this be simplified to `if (--lifetime <= 0 && !force)`?

File: OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs (hunk 3 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs]: Fixed.

File: OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs (hunk 4 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs]: Fixed.

File: OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs (hunk 5 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs]: Fixed.

**Review by teinarss — DISMISSED** [review]
looks good

**Comment by pchote** [comment]
I still haven't had time to look at the technical aspects of it, but I stand by my earlier concerns that the change to the target line thickness and colours will be controversial and risks generating drama that could potentially overshadow all the great things that have come from this rework. I will be much happier if we can split the colour and thickness changes into a separate PR to target Next + 1, and I don't mind doing this myself if needs be.

**Comment by tovl** [comment]
The color changes have already been reverted (except for attack-move, which I still believe is important to retain information).

The line thickness has also been changed to 2 px with 3 px endnodes, which, judging by the discussion in #16616, seems like a reasonable compromise until a more advanced solution is implemented.

**Comment by abcdefg30** [comment]
Needs a rebase, btw.

**Comment by tovl** [comment]
Rebased.

**Comment by pchote** [comment]
> (except for attack-move, which I still believe is important to retain information).
Oh 🙁

This was actually the main thing that I considered to be a deal breaker with the original changes. For years we have used yellow as a special case to mean "guard unit" / "repair building", and I find having it change to being one of the most common order line colors jarring and offputting.

File: OpenRA.Mods.Common/Activities/Move/Move.cs [hunk 6]
```diff
@@ -57,10 +58,12 @@ public Move(Actor self, CPos destination)
 				return path;
 			};
 			this.destination = destination;
+			this.targetLineColor = targetLineColor;
 			nearEnough = WDist.Zero;
 		}
 
-		public Move(Actor self, CPos destination, WDist nearEnough, Actor ignoreActor = null, bool evaluateNearestMovableCell = false)
+		public Move(Actor self, CPos destination, WDist nearEnough, Actor ignoreActor = null, bool evaluateNearestMovableCell = false,
+					Color? targetLineColor = null)
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Activities/Move/Move.cs]: That's two tabs more than we usually use in such cases.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 7]
```diff
@@ -1051,9 +1048,8 @@ void Nudge(Actor self)
 				.Rotate(WRot.FromFacing(self.World.SharedRandom.Next(256)));
 			var target = Target.FromPos(self.CenterPosition + offset);
 
-			self.CancelActivity();
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Intentional? If yes, I think this would be better in a separate commit with a short explanation, as this doesn't really have to do with the target lines.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 8]
```diff
@@ -1051,9 +1048,8 @@ void Nudge(Actor self)
...
-			self.SetTargetLine(target, Color.Green, false);
-			self.QueueActivity(new Fly(self, target));
+			self.QueueActivity(true, new Fly(self, target));
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Same here, also shouldn't this be `false` then? Using that QueueActivity overload with `true` basically just calls `QueueActivity(Activity)`, making the `true` redundant.

**Comment by reaperrr** [comment]
> Having the rally point line width not matching the target lines is also not ideal.

I agree.

> This was actually the main thing that I considered to be a deal breaker with the original changes. For years we have used yellow as a special case to mean "guard unit" / "repair building", and I find the change to it now being one of the most common order line colors jarring and offputting. I suspect that I won't be the only person to have this reaction.

In my opinion, the target line color should match the cursor color, which is why I agree it should stay red (at least for now).

**Comment by tovl** [comment]
Ok, let me explain my reasoning in full:

First off, "guard" and "repair building" represent two completely different actions, so there is no particular reason why they should have the same color. The only real reason to give them the same color is that they both are situational (so not as likely to be confused as some other combinations) and to limit the palette of different line colors somewhat. The requirement to keep the number of colors limited (which I do ascribe to) does not prescribe a specific number; It really depends on our other requirements. All in all, we can be rather flexible here.

 "guard" and "attack-move" are almost the same thing. Both in concept and in terms of effect: "guard" is simply "attack-move" with a moving endnode. Furthermore, they are fully complementary: All valid targets for "guard" are invalid targets for "attack-move" and vice versa. It makes sense then to give them the same color. 

The fact that they are complementary also makes it actually possible to unify them UI wise without losing any granularity of control. This is something that I would like to at least try to see how that will be received for next+1 in the context of #16669 and #16381. My reason for wanting to condense these order modifiers is that it would allow us to implement things like #6167 and/or #4968 without cluttering the hotkeys/commandbar interface too much. Since there is no loss of granularity of control, this is a win-win in my book, but it would be be a more radical change than a simple color switch. 

Finally, the fact that "attack-move" is one of the most common orders is precisely why it is so important for it to be visually distinct from the other two most common orders: "move" and "attack". Especially now that it is possible to form long chains of orders, and since the entire point of this PR is to make those chains visible, you want to be able to know which line segment belongs to what order. We can gloss over the more esoteric commands to a degree, but separating the most common and conceptually fundamental order types is essential to the purpose of this PR.

In summary, there are only two hard requirements here:
- "guard" and "attack-move" need to have the same color
- "attack-move" needs to have a different color from both "move" and "attack"

- optionally, give "repair" and other situational commands a fourth color, but this needs to be weighed against the demand to minimize the number of distinct colors.

Note that in the entire preceding wall of text I have not mentioned any specific color even once. I have no strong opinions about them (although I thought the traffic light analogy for "move" -> "attack-move" -> "attack" was rather cute). That means there is actually a lot of room to accommodate your concerns, but you need to give me something to work with. 

There are alternatives that do not touch the "specialness" of yellow. Orange for "attack-move"/"guard" was actually my first idea, but I found it had very poor contrast in desert maps. Maybe simply a different shade of red or purplish-red would reduce the "jarringness" because it wouldn't look too different from before but still be distinguishable to some degree?

> Having the rally point line width not matching the target lines is also not ideal.

It would not be a lot of effort to change them to match.

**Comment by pchote** [comment]
I can't argue against any of your logic there, and admit that there is probably a large amount of knee-jerking in my reaction to the colour change. I'll try poking around with some different color variations and see if there is another option that feels better there.

**Comment by Punsho** [comment]
The of the bigger problems is that we can only choose from 3 colours, green, yellow, red. Others are either hard too see in clutter of colours or do not fit with the RA/TD style. Here's the list of bright alternatives I got from looking at various palettes: white, orange, light blue, pink, light purple, light brown. 

> Maybe simply a different shade of red or purplish-red would reduce the "jarringness" because it wouldn't look too different from before but still be distinguishable to some degree?

I don't think that we should have darker variants of the colours we currently have. They're just hard to distinguish from background noise

**Comment by obrakmann** [comment]
> "guard" and "attack-move" are almost the same thing. Both in concept and in terms of effect: "guard" is simply "attack-move" with a moving endnode.

If I may butt in from the sidelines here and wax a bit philosophical: I think the quote above probably illustrates the source of your disagreement. @tovl is looking at the problem from the inside out, while @pchote is probably looking outside-in. And if Paul would do some soul-searching, he would probably also realise that his view of what the Guard feature is supposed to do doesn't quite match with the actual implementation.

Imagine you wouldn't know anything OpenRA's internals. How would you then implement a Guard feature?  I can think of at least four ways that may in part be combined:

1. OpenRA's way: Attack-Move towards the guarded unit's location
2. Follow the guarded unit. Use the guarded unit to scan for targets and hand them off to the guard.
3. Follow the guarded unit. Let the guarded unit inform the guard when it gets attacked and hand off the attacker as a target.
4. Drive ahead of the guarded unit, attack-move towards its destination. Guarded unit follows guard.

OpenRA's current way is the most simple of those ways, and the others are all quite different to a simple attack-move (and thus warranting a different colour, IMHO), with the last one even kind-of reversing the relationship.

NB: The solution 2 above would also allow for interesting applications wrt AI, commander units, intelligent scouts and so on.

**Comment by tovl** [comment]
> Imagine you wouldn't know anything OpenRA's internals. How would you then implement a Guard feature?

I wouldn't. I would implement unit formations instead.

I think we are actually all in agreement about what we wish the hypothetical "guard" feature (let's just call it "escort" for the sake of argument) to be in the abstract. The only difference is that I have already drawn the conclusion for myself that the current "guard" feature is never going to satisfactorily fill that "escort" role no matter how much we mangle it.

Your option number 4 comes closest to what it needs to be, because what an "escort" really needs to do is drive ahead and around the vulnerable units in a screening formation. However, without actual formation logic this is not going to work, because we need to define sensible behaviour for when multiple units are escorting the same target. Number 4 does not work at all with multiple escorts.

So, accepting that we would need a completely fresh and separate mechanism to have a true "escort" feature anyway, it then makes sense to just lean into what "guard" actually is: A minor extension on the "attack-move" feature. You hit the nail on the head by saying I have an inside-out perspective on it, and this is why.

As an aside, keep in mind that the current "guard" feature is exactly what it was in the original games. There is nothing wrong with just keeping it that way (and possibly folding into "attack-move" without loss) and taking our time to come up with a new feature.

**Comment by GraionDilach** [comment]
If I'd bump in from the sidelines - number 2 was used by Warzone 2100's commander tanks as a separate feature of them.

File: OpenRA.Mods.Common/Activities/Move/Move.cs (hunk 6 above)
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Move/Move.cs]: Fixed.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 8 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: This should be `false` of course, which makes it a very minor cleanup.

File: OpenRA.Game/Graphics/TargetLineRenderable.cs [hunk 9]
```diff
@@ -42,22 +46,22 @@ public void Render(WorldRenderer wr)
 			if (!waypoints.Any())
 				return;
 
-			var iz = 1 / wr.Viewport.Zoom;
+			var iz = width / wr.Viewport.Zoom;
```
Comment (pchote) [inline OpenRA.Game/Graphics/TargetLineRenderable.cs]: nit: `iz` was short for "inverse zoom", so should be renamed now that this is a scaled width.

File: OpenRA.Mods.Cnc/Traits/PortableChrono.cs [hunk 10]
```diff
@@ -114,12 +114,12 @@ public void ResolveOrder(Actor self, Order order)
 					self.CancelActivity();
 
 				var cell = self.World.Map.CellContaining(order.Target.CenterPosition);
-				self.SetTargetLine(order.Target, Color.LawnGreen);
 				if (maxDistance != null)
-					self.QueueActivity(move.MoveWithinRange(order.Target, WDist.FromCells(maxDistance.Value)));
+					self.QueueActivity(move.MoveWithinRange(order.Target, WDist.FromCells(maxDistance.Value), null, Color.LawnGreen));
```
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/PortableChrono.cs]: If you specify this as `targetLineColor: Color.LawnGreen` you can skip the `null`, and make this more consistent with other activities.

File: OpenRA.Mods.Common/Activities/Air/Land.cs [hunk 11]
```diff
@@ -251,5 +254,10 @@ public override bool Tick(Actor self)
 
 			return false;
 		}
+
+		public override IEnumerable<TargetLineNode> TargetLineNodes(Actor self)
+		{
+			yield return new TargetLineNode(target, targetLineColor.Value);
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/Land.cs]: Shouldn't this be checking `targetLineColor != null` first?

File: OpenRA.Mods.Common/Activities/PickupUnit.cs [hunk 12]
```diff
@@ -110,6 +111,11 @@ public override bool Tick(Actor self)
 			return true;
 		}
 
+		public override IEnumerable<TargetLineNode> TargetLineNodes(Actor self)
+		{
+			yield return new TargetLineNode(Target.FromActor(cargo), Color.Green);
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/PickupUnit.cs]: Bleed uses yellow here.

File: OpenRA.Mods.Common/Activities/Turn.cs [hunk 13]
```diff
@@ -43,5 +45,10 @@ public override bool Tick(Actor self)
 
 			return false;
 		}
+
+		public override IEnumerable<TargetLineNode> TargetLineNodes(Actor self)
+		{
+			yield return new TargetLineNode(Target.Invalid, Color.Gray);
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Turn.cs]: Is this one needed at all?

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 14]
```diff
@@ -972,8 +972,8 @@ public void ResolveOrder(Actor self, Order order)
 					UnReserve();
 
 				var target = Target.FromCell(self.World, cell);
-				self.SetTargetLine(target, Color.Green);
-				self.QueueActivity(order.Queued, new Fly(self, target));
+				self.QueueActivity(order.Queued, new Fly(self, target, null, Color.Green));
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Style nit: another case where we can do `targetLineColor: Color.Green` and skip the `null`.

File: OpenRA.Mods.Common/Traits/Attack/AttackBase.cs [hunk 15]
```diff
@@ -214,7 +214,7 @@ string IOrderVoice.VoicePhraseForOrder(Actor self, Order order)
 			return order.OrderString == attackOrderName || order.OrderString == forceAttackOrderName ? Info.Voice : null;
 		}
 
-		public abstract Activity GetAttackActivity(Actor self, Target newTarget, bool allowMove, bool forceAttack);
+		public abstract Activity GetAttackActivity(Actor self, Target newTarget, bool allowMove, bool forceAttack, bool ordered = false);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]: Can you elaborate on the need for the `ordered` flag?
Plumbing this through to the activities so they can suppress their target lines makes up a significant fraction of the changes here, so i'd like to make sure there isn't a more concise solution.

It looks like this is mainly for `AttackMove`, but can't we solve that at the source by having that simply not chain through to the child attack activities?

File: OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs [hunk 16]
```diff
@@ -62,65 +45,61 @@ void INotifySelected.Selected(Actor a)
...
-				if (target.Type == TargetType.Invalid)
+				if (a.IsCanceling)
 					continue;
 
-				yield return new TargetLineRenderable(new[] { self.CenterPosition, target.CenterPosition }, c);
+				foreach (var n in a.TargetLineNodes(self))
+				{
+					if (n.Target.Type != TargetType.Invalid)
+					{
+						yield return new TargetLineRenderable(new[] { prev, n.Target.CenterPosition }, n.Color, 2, 3);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs]: Can we please expose the `LineWidth` and `MarkerWidth` on the `DrawLineToTargetInfo`?

File: OpenRA.Game/Graphics/TargetLineRenderable.cs (hunk 9 above)
Comment (tovl) [reply OpenRA.Game/Graphics/TargetLineRenderable.cs]: Fixed.

File: OpenRA.Mods.Cnc/Traits/PortableChrono.cs (hunk 10 above)
Comment (tovl) [reply OpenRA.Mods.Cnc/Traits/PortableChrono.cs]: Fixed.

File: OpenRA.Mods.Common/Activities/Air/Land.cs (hunk 11 above)
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/Land.cs]: Fixed.

File: OpenRA.Mods.Common/Activities/PickupUnit.cs (hunk 12 above)
Comment (tovl) [reply OpenRA.Mods.Common/Activities/PickupUnit.cs]: Fixed.

File: OpenRA.Mods.Common/Activities/Turn.cs (hunk 13 above)
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Turn.cs]: Nope. Removed.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 14 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Fixed.

File: OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs (hunk 16 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs]: Done.

File: OpenRA.Mods.Common/Traits/Attack/AttackBase.cs (hunk 15 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]: `AttackMoveActivity` already doesn't chain through to child activities (like most activities) so it isn't needed for that. It is to prevent `AutoTarget` from drawing target lines. 

The reasoning being: Opportunity attacks during a regular move don't draw target lines, automatic attacks during attack-move don't draw targeting lines, so it would be consistent if automatic attacks while stationary also don't draw target lines. Only when an attack is explicitly ordered does it draw a target line.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]: Ok, that makes sense and I can't think of a better alternative at this point. Can we please split this into its own commit?

Comment (pchote) [reply OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]: We may want to show the target lines if `!ordered` but `allowMove` because it would look weird for the unit to start chasing the target around the map without having a target line.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]: Can we please replace ordered with a targetLineColor? This would make the attack activities consistent with the others that allow the caller to override the line behaviour, and would let us give e.g. heal weapons a different line colour in a follow up pr.

Comment (tovl) [reply OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]: about the `allowMove` part:

I admit it would be awkward to not know what the unit is chasing there. On the other hand, this is the same situation as with `attackMove` so it would be consistent. It is also a bit awkward to see order lines appearing when you didn't order anything.

In the long term, as I indicated in the OP, I would like to reintroduce target lines as a completely separate thing from order lines. It would be a 'weaker' (thinner and/or dashed and/or transparent) line, called from the trait code and pointing to whatever is the current (opportunity) attack target. But this will have to wait for the `Attack*` refactor. 

I'm not sure what the best option for the short term is though.

Completely agree on the `targetLineColor` part.

Comment (tovl) [reply OpenRA.Mods.Common/Traits/Attack/AttackBase.cs]: Replaced `ordered` with `targetLineColor` and made the attack colour yaml configurable.

**Comment by tovl** [comment]
Rebased.

**Comment by tovl** [comment]
Implemented possible alternative colour-scheme that keeps yellow exclusive to special actions:
`OrangeRed` for guard/attack-move and `Crimson` for regular attacks and other offensive actions.

`OrangeRed` gives a better contrast in arid maps than `Orange.` Changing from `Red` to `Crimson` gives more color separation with `OrangeRed.` Both keep reasonably close to `Red` so shouldn't be too jarring compared to `Yellow.`

Screenshot from 2019-07-23 21-53-29
Screenshot from 2019-07-23 21-46-25

**Comment by pchote** [comment]
I have pushed several fixups directly to the branch with a goal of avoiding another round of delays:
- Added a descriptive commit message
- Fixed the missing target lines for TD gunboats (which can be given orders in missions)
- Removed unused usings from Turn
- Split color changes into their own commit for our future sanity when looking back at what and when the colors changed

Unfortunately, while implementing those I noticed a couple of other points that IMO need further discussion:
* Harvesters when left to themselves don't show any target lines, and when explicitly ordered to unload show a permanent target line to their linked refinery. This is either a bug or, if intentional, inconsistent with how target lines work for everything else.
* IMO we should consider removing the target line display from alt, and have it only on shift. We now use alt for various thinks that aren't queuing related, whereas shift is always queuing.
* The guard cursor is yellow, so having redish target lines still feels very wrong for me, even if it is behaving the same as attack-move at a code level (which we should fix). The other attack color changes LGTM.

**Comment by tovl** [comment]
> Harvesters when left to themselves don't show any target lines, and when explicitly ordered to unload show a permanent target line to their linked refinery. 

Fixed by deferring to childactivities when appropriate.

> we should consider removing the target line display from alt, and have it only on shift. 

I agree, but might be controversial. Removed it tentatively, pending further feedback. I do think it would be a good idea to introduce a separate hotkey to toggle target lines on/off permanently in addition to shift, but that is something for the future.

> The guard cursor is yellow,

Actually, that is not quite true. In TS it is red (with green corner pips and above it a silver shield with gold detail, but predominantly red), in D2K it is blue (!) and in RA (okay this one is a bit nitpicky) it is gold/brown (`GoldenRod` is the corresponding html color). Only in TD is it actually yellow.

Also if fidelity to the cursors is the measure of things then we should definitely make harvest red (not green), pick-up green (not yellow) and repair-building should in TS actually also be green. (In fact, these would all be quite reasonable changes in my view)

For the future it might be a good idea to make all these line colors configurable via yaml anyway so that it can be adapted on a per mod basis (my main reason for wanting this has more to do with terrain background colors). But I don't want to delay this further over that and for now reddish orange actually fits very well with TS's cursor. On RA I guess we'll just have to agree to disagree because I don't find it jarring with RA's gold at all.

**Comment by pchote** [comment]
Some more target line bugs:

* Harvesters that have an unload order with a harvest order queued afterwards don't show the target line to the refinery while driving back to it (it appears once it starts docking)
* Minelayers don't show target lines while they are mining a field - they instead show a line straight to whatever is queued afterwards
* MAD Tank and MCV deploying doesn't act as a terminal activity (probably depends on #16805)
* Aircraft (maybe restrict to `AbortOnResupply: true` if it makes things simpler) keep their attack line on their target while they are returning to reload, and while reloading - but then drop their target and the line once reloading is complete. It would feel more consistent if the target line changed to the airfield immediately on return.

**Comment by tovl** [comment]
> Harvesters that have an unload order with a harvest order queued afterwards don't show the target line to the refinery while driving back to it (it appears once it starts docking)

Fixed.

>Passengers entering a transport doesn't act as a terminal activity

Fixed.

> Aircraft (maybe restrict to AbortOnResupply: true if it makes things simpler) keep their attack line on their target while they are returning to reload, and while reloading - but then drop their target and the line once reloading is complete. It would feel more consistent if the target line changed to the airfield immediately on return.

Fixed. With `AbortOnResupply: true` the target line is replaced on return; With `AbortOnResupply: false` the targetline to the airfield is inserted before the attack line on return.

MCV deploying acting like a terminal activity is really tricky, because it would have to depend on the future deploy state of the MCV. This is also related to queued deploy orders not registering while the MCV is still undeploying. I also noticed that move-undeploy has regressed again.

**Comment by pchote** [comment]
Hmm, that is a very good point about not knowing whether the MCV deploy will actually work. The same applies for passengers too - the transport may be full. It is probably best to keep the remaining target lines for all of these cases?

**Comment by tovl** [comment]
Ok. Removed `IsTerminal` entirely.

Also fixed minelayer not showing target lines at all and hooked the minefield visualization into the new target line logic because that thing was completely broken (see #15726).

It would look even  better if #15727 was implemented as well but that would touch on the actual unit behaviour and that doesn't seem appropriate for this PR.

**Comment by pchote** [comment]
I'm sorry to say that harvesters still aren't quite right :cry:

I'm not sure exactly how to reproduce it, but it happened when I had several harvest ordered queued, and then a move order. The harvester returned to unload with I think one harvest order left in the queue followed by the move. While it was returning the target line showed as harv -> refinery -> queued move location instead of harv -> refinery -> queued harvest location -> queued move location.

I'll be mostly afk from ~now until next monday, so hopefully someone else can come up with a more firm repro case here.

**Comment by pchote** [comment]
Re minelayers: would it be too much to sneak something like 42d7ae29fe611d41941dbc519ed54c3d24911339 in here?

**Comment by pchote** [comment]
This is generally looking good to me, so we may want to merge this sooner rather than later and push any remaining bug fixes to followup PRs.

**Comment by tovl** [comment]
> would it be too much to sneak something like 42d7ae2 in here?

That looks like a good change, but considering your next remark it seems more prudent to me to make that a follow-up PR.

> This is generally looking good to me, so we may want to merge this sooner rather than later and push any remaining bug fixes to followup PRs as well.

I think that would be a good idea. This is probably the last big-ticket feature for next release, so with this out of the way we can concentrate on the regression whack-a-mole. It would be a good idea to make an issue for the harvester bug then; It might take a while to figure out such a hard to replicate bug.

**Comment by teinarss** [comment]
Think that i can reproduce it now, queue 5 harvest orders and one move.
harv_waypoint

**Comment by Punsho** [comment]
@teinarss you can hold `alt` to keep those lines visible

**Comment by tovl** [comment]
The harvester bug should be fixed now.

**Review by teinarss — DISMISSED** [review]
Good work!

**Comment by abcdefg30** [comment]
Needs a rebase.

**Comment by tovl** [comment]
Rebased.

**Review by pchote — APPROVED** [review]
The one remaining big issue I found was that this disables the target lines for spectators / ally players.

I have a PR lined up to bring this back with some additional tweaks, so no need to block this further.
Followup refinement and fixes can/will come in followup PRs.

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/0868d274e31f3e46820aa4d6321caf5b8b1f6686

**Comment by Turupawn** [comment]
Awesome guys! Congrats! :tada: :tada:
